### DISCUSSION TITLE
Add file to existing archive

### DISCUSSION THREAD

**Discussion by jhammer** [maennchen/ZipStream-PHP#261] [Ideas]
Hi. Love ZipStream. We have a use case where we’d like to read a ZIP archive that is already on disk and stream it out to the user, but add file(s) to the stream as it goes out. The original ZIP archive on disk should remain untouched. We do not need to read or modify any of the files that are already in the stream.

The `ZipStream` constructor could be modified to take an optional `inputPath` or `inputStream` parameter. No other public API changes would be necessary.

We would be happy to sponsor development of this feature if you think it is feasible.

**Comment by maennchen** [comment]
Hmm. I think it should be possible.

I think appending files is a lot easier since the existing file headers do not have to be recomputed.

To do so, we‘d have to be able to parse header entries and skip existing files in the inputStream. As soon as we reach the central directory header, we‘d have to parse it, record its contents, write our own files and then emmit a new merged end on central directory record.

Another possibility to an inputStream would be a addFilesFromZip[Stream|Path] function. That would allow adding multiple zip contents.

The thing I‘m most worried about with this feature is that we‘d potentially read zip files which are using extensions / functionalities that this library does not support that might cause problems. We‘d probably have to be quite strict in what we parse initially.

One more option would be to use PHPs internal ZipArchive class and just loop its contents. But I don’t think that is what you were looking for.

I‘ll convert this to an issue to discuss the details of a possible implementation.

**Reply by maennchen** [reply]
Moved to #262
